psydac/feec/multipatch/examples/ppc_test_cases.py

```diff
@@ -62,23 +254,171 @@ def get_source_and_sol_for_magnetostatic_pbm(
+        f_x =   dy_tau * phi
+        f_y = - dx_tau * phi
+        f_vect = Tuple(f_x, f_y)
 
-def get_source_and_solution(source_type=None, eta=0, mu=0, nu=0,
+    else:
+        raise ValueError(source_type)
+
+    # u_ex = Tuple(0, 1)  # DEBUG
+    return f_vect, u_bc, u_ex, curl_u_ex, div_u_ex #, phi, grad_phi
+
+def get_source_and_solution_h1(source_type=None, eta=0, mu=0,
+                            domain=None, domain_name=None):
+    """
+    provide source, and exact solutions when available, for:
+    
+    Find u in H^1, such that
+
+      A u = f             on \Omega
+        u = u_bc          on \partial \Omega
+
+    with
+
+      A u := eta * u  -  mu * div grad u
+    
+    see solve_h1_source_pbm()
+    """
+
+    # exact solutions (if available)
+    u_ex = None
+
+    # bc solution: describe the bc on boundary. Inside domain, values should not matter. Homogeneous bc will be used if None
+    u_bc = None
+
+    # source terms
+    f_scal = None
+
+    # auxiliary term (for more diagnostics)
+    grad_phi = None
+    phi = None
+
+    x,y    = domain.coordinates
+
+    if source_type in ['manu_poisson_elliptic']:
+        x0 = 1.5
+        y0 = 1.5
+        s  = (x-x0) - (y-y0)
+        t  = (x-x0) + (y-y0)
+        a = (1/1.9)**2
+        b = (1/1.2)**2
+        sigma2 = 0.0121
+        tau = a*s**2 + b*t**2 - 1
+        phi = exp(-tau**2/(2*sigma2))
+        dx_tau = 2*( a*s + b*t)
+        dy_tau = 2*(-a*s + b*t)
+        dxx_tau = 2*(a + b)
+        dyy_tau = 2*(a + b)
+
+        dx_phi = (-tau*dx_tau/sigma2)*phi
+        dy_phi = (-tau*dy_tau/sigma2)*phi
+        grad_phi = Tuple(dx_phi, dy_phi)
+
+        f_scal = -( (tau*dx_tau/sigma2)**2 - (tau*dxx_tau + dx_tau**2)/sigma2
+                   +(tau*dy_tau/sigma2)**2 - (tau*dyy_tau + dy_tau**2)/sigma2 )*phi
+
+        # exact solution of  -p'' = f  with hom. bc's on pretzel domain
+        if mu == 1 and eta == 0:
+            u_ex = phi
+        else:
+            print('WARNING (54375385643): exact solution not available in this case!') 
+
+        if not domain_name in ['pretzel', 'pretzel_f']:
+            # we may have non-hom bc's            
+            u_bc = u_ex
+
+    elif source_type == 'manu_poisson_2':
+        f_scal = -4
+        if mu == 1 and eta == 0:
+            u_ex = x**2+y**2
+        else:
+            raise NotImplementedError
+        u_bc   = u_ex
+
+    elif source_type == 'manu_poisson_sincos':
+        u_ex    = sin(pi*x)*cos(pi*y)
+        f_scal  = (eta + 2*mu*pi**2) * u_ex
+        u_bc = u_ex
+
+    else:
+        raise ValueError(source_type)
+
+    return f_scal, u_bc, u_ex
+
+
+
+def get_source_and_solution_OBSOLETE(source_type=None, eta=0, mu=0, nu=0,
                             domain=None, domain_name=None,
                             refsol_params=None):
     """
-    compute source and reference solution (exact, or reference values) when possible, depending on the source_type
+    OBSOLETE: kept for some test-cases
     """
```

Comment:
Is there a way to remove this function completely?

Reply:
See https://github.com/pyccel/psydac/pull/320#discussion_r1616276346.